### ISSUE TITLE
Imperative Migrations with a TypeScript DSL

### ISSUE THREAD

**Issue by flybayer** [prisma/prisma#4688]
## Problem

Not all migrations can be declarative. For example, we need to add a custom index for performance reasons. We also need this as an escape hatch for anything prisma schema doesn't support. Lastly, we also need this for data migrations.

This is very important for any serious apps using prisma migrate.

## Suggested solution

A new command: `prisma migrate create`

This will create the necessary migration files. Similar as `prisma migrate save` but the files are just stubbed out.

Then the user will open the stubbed migration file and add their custom migration code.

Perhaps the imperative migration file can look something like this:

```ts
import {migrate} from '@prisma/migrate`

migrate({
  async up(prisma) {
    // SQL migration
    await prisma.executeRaw(`some SQL stuff`)

    // Data migration
    await prisma.someModel.create(/*..*/)
  },

  async down(prisma) {
    await prisma.someModel.delete(/*..*/)

    await prisma.executeRaw(`undo some SQL stuff`)
  }
})
```

**Comment by machendos** [comment]
@albertoperdomo are you currently working on migrations using DSL of typescript? Now, as for me, the absence of this possibility is the only thing that stops from introducing the prism to large projects. Since without it we can't applying to DB changed, based on the typescript, for example, inserting values from dictionary enums in automatic mode. Or, was the work on migrations finished after the introduction of migrations based on the SQL?

**Comment by florianmartens** [comment]
I'd also love some info on this. The fact that this seems to be the way Prisma wants to move forward could be a real concern / show-stopper for me. While it might be sufficient for small projects to use migrations that are .sql only, this introduces pretty drastic contrains for larger projects.

It would be great to know if the Prisma team conciders .sql migrations as sufficient or wants to move to a solution that allows for scripting.

**Comment by tomhoule** [comment]
It's a complex question, there is no written stance at Prisma at the moment, the discussion is still definitely open. This issue was opened at the time of the experimental migrate version, where migrations were `.json` files, declarative, non-deterministic and with very little customization possible. The migrate we have now, with SQL migrations, gives you a lot more freedom, while still being deterministic. It was also the easiest to implement reliably and without spending time and effort designing a DSL (like any other software project, we have to prioritize between many features and we have limited time).

My individual, personal view is that since the SQL migrations are generated, you only need to review them and maybe customize them, which is easier than writing them yourself from scratch, so I think they are fine.

These are a few questions that come to mind when discussing a Typescript DSL for migrations on SQL databases:

- What is the execution model of the DSL? Is it a script that results in an object that will be interpreted to SQL, or a script that can interact directly with the database, do API calls, use dependencies, etc.? The first case is a lot more restricted, but would give Prisma more context on the content of the migration, and can be made deterministic. The second case introduces much more complexity, and lets you write migrations that will easily break when you upgrade your TS dependencies. Migrations can't easily be refactored, they have to keep working the same for as long as your project keeps them around.

- Migrate fundamentally relies on migrations being determinisitc. Applying the same sequence of migrations twice should always result in exactly the same database schema, otherwise we can't reliably generate new migrations and even diagnose what the database schema actually is/should be. This means authors of TS migrations would have to be very careful with conditional logic and non-deterministic things like API calls.

**Comment by florianmartens** [comment]
Thanks @tomhoule, what you say definitely makes sense and is very interesting to think about. My current use case is as follows. I'm working with a simple text based field named `description`. We now decided to let users interact with this field using a WYSIWYG editor and need to transform existing `description`s into a stringified version of an WYSIWYG competible object. The transformation requires the use of a library.

This is a relatively simple use-case but around 80% of my migration scripting use-cases are in that ballpark. Both migrations hooks as well as DSL migrations would potentially solve this (although I'd prefer DSL migrations).

I understand the need for deterministic migrations but don't think this necessarly needs to be enforced programatically if it leads to limitations as such. What is wrong with advising users to not apply db schema migrations conditionally? I think many of us have experience with other migrations systems that wouldn't stop us from doing this, yet I'm assuming that very few of us have ever mutated our db schema conditionally as it screams bad practise.

I also understand that your resources are limited and am super thankful for all the work you've put in already. I just think prisma would really become so much more useful if the migration system would support:
- Down migrations
- Scripting
- Migrating to an arbitrary point in the migration history

**Comment by tomhoule** [comment]
Thanks for the context @florianmartena, that definitely makes a lot of sense. You raised the idea that migration hooks would also be OK for that, and I think we're more aligned than it looks like at first. If I understand your use case correctly, it's about using a JS/TS library to migrate _data_, rather than the database schema.

There is room for improvement in migrate to support data migrations, and there a DSL would totally be the way to go — the question is whether that DSL should also be used for schema migrations. There is a lot to be said for keeping them separate (see [this classic blog post](https://thoughtbot.com/blog/data-migrations-in-rails) from the Rails world. If you have input for designs in that direction, it is super welcome and will definitely be taken into account — there is a lot of feedback in GitHub issues and a lot of different projects running at the same time, but we always read and discuss all of it. This problem could also benefit from its own issue.

**Comment by vladinator1000** [comment]
Hi y'all, what about generating imperative .sql migrations with code? This way only the generated .sql needs to be correct and it can use the existing migration history system.

**Comment by Jolg42** [comment]
Is there actually any reason why Prisma doesn't have the same migration style as TypeORM, which just generates typescript/js migration classes with async `up` and `down` functions? That way migrations are run with access to the code and from within the code, which seems like the more logical way of doing things to me. 

Apart from the issue with serverless as others pointed out, I honestly can't imagine how I would use prisma for some more complex migrations that also require business logic to run over the data, which isn't a big deal when using TypeORM as you can just do whatever you want in the code while running the migration. Sometimes I can imagine you might also need to connect to an external API to manipulate the data while migrating and that's just something you will very hardly do with plain SQL.

I am now in process of deciding whether to use Prisma or TypeORM for the next project and this seems like a deal-breaker honestly, although in the rest of the features Prisma seems like a clear winner.

_Originally posted by @vnenkpet

**Comment by hyunwoona** [comment]
Any updates on this? I desperately need this feature and there's no updates for months

**Comment by franky47** [comment]
One issue I see with using TypeScript to run data migrations is the schema evolution over time.

Schema changes like renaming or removing fields or entire models would break previous data migration code, since the Prisma client only exists in its current version.

**Comment by LMaxence** [comment]
> One issue I see with using TypeScript to run data migrations is the schema evolution over time.
> 
> Schema changes like renaming or removing fields or entire models would break previous data migration code, since the Prisma client only exists in its current version.

From my perspective, what should be provided in such an API is not the same functionalities as the client provides. What could be easy, and stable to provide, would be an API able to manipulate the table structure. [For instance, Laravel does it well](https://laravel.com/docs/9.x/migrations#creating-tables).

The main idea behind this is to write database migrations, in the end doing the same job as the SQL migration files. However, writing them in some nice scripting language would be more efficient for migrating the data once the structure has evolved. I don't recall the number of hours I have left so far manipulating JSONb columns in these SQL files.

**Comment by DavidVaness** [comment]
I can only voice my support for @LMaxence point and would concur that Laravel's way of handling this is really helpful.

**Comment by emmanuelbuah** [comment]
I agree with the example provided by @flybayer . Kysely does exactly that - https://github.com/koskimas/kysely#postgresql-migration-example

**Comment by franky47** [comment]
@LMaxence I was more referring to "data migrations": once the schema is in a given state (changed either via the current `prisma migrate` SQL-based approach or by a code-driven approach as you propose), running code at runtime to modify records based on some logic.

This is often used in [expand-contract migrations](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern), where for example you would want to add a new column which contains data that can be computed from a mix of the existing record and external data:

1. Add the new column, marked as optional
2. Run data migrations to iterate over records and fill up the new column with computed data
3. Mark the new column as required and remove old columns (for normalisation)

If step 2 was to use the Prisma Client to run those data migrations, then it would be hard to keep it in sync across schema versions. But because this is a TypeScript matter only, and those data migrations should only run once like their SQL counterparts, a simple `// @ts-nocheck` at the top of the data migration file would work to suppress those errors.

**Comment by vnenkpet** [comment]
@franky47 I don't even think using prisma client as-is (with all the schema type checking) is necessary. It would actually be totally OK for me to use some sort of DSL/raw SQL queries inside the code to handle what I need, the issue most of us have with this (I assume) is that that's impossible and we're forced to rely on plain SQL files to do complex migrations, which, at least for me, kind of defeats one of the main reasons why I use ORM's in the first place.

Prisma so far is great in almost everything I tried and this is one of the blockers why I worry switching to it completely on production projects because I simply don't see a simple way how to handle bigger migrations involving data manipulation efficiently without having to re-learn some complex SQL operations that I feel should be abstracted/handled by TS code.

**Comment by SimonPiscitelli** [comment]
Our team is migration back to Knex from migrations.

Please make Prisma migrations like Knex or Eloquent or Doctrine

**Comment by camsloanftc** [comment]
Would it be possible / make sense to take a snapshot of the prisma schema when migrations are executed, and save in the migration folder? With this, there could be some way to reference `previous` / `next` schema as you run an imperative migration. Maybe that's too much to manage in terms of typescript types generation though.

If that sounds totally outlandish, I would still be very much in favour of having some way to run imperative data migrations without type safety and type autocompletion, like the previous commenters have suggested.

**Comment by LMaxence** [comment]
# TL;DR

From my own user perspective, and for a very basic first version of the feature we are discussing:

- I don't really care about typings for this
- I don't really care about using the prisma client for this
- I just would like to use JS/TS/whatever nice scripting language code to write modifications of my data, regardless of this data being typed or not
- I'm completely fine with writing raw SQL queries to do the part "fetching the existing data", and "saving the migrated data"

# Detailed point of view

I completely understand the problem with using the client to manipulate data across migrations, as typings may change over time and invalidate existing migrations.

That's why I suggested a first basic approach of the issue, where this feature would have nothing to do with the Prisma client. The API we are talking about can be really simple and only go down to generated SQL statements. For instance:

```SQL
ALTER TABLE "users" ADD COLUMN "something_new" VARCHAR NOT NULL;
```

could go down to

```
async up() {
   db.table("users").addColumn("something_new", {type: PSQLTypes.VARCHAR, notNull: true})
}
```

I agree this is not a big improvement, but that can become one once I can do stuff like:

```ts
async up() {
   await db.table("myEntity").addColumn("somethingPreiouslyStoredInAJsonbCol")
   const rows = await db.table("myEntity").select("id", "myHorribleJSONbColumn").rows()
   // I'm 100% fine with the two lines above being `$executeRaw` statements
   
   for (row of rows) {
       // This is what takes hours to do when manipulating complex data with SQL statements
       const { removedAttribute, ...remainingJSONb } = row.myHorribleJSONbColumn
       // I'm 100% fine with this line below being an `$executeRaw` statement
       await db.table("myEntity").update({
           set: {somethingPreiouslyStoredInAJsonbCol: removedAttribute, myHorribleJSONbColumn: remainingJSONb}, 
           where: {id: row.id}
       })
   }

}
```

The problem is not about writing structure migrations in SQL, I mean it's totally fine to write `ALTER TABLE` statements, or `UPDATE` statements for scalar columns. However, things always get messy once I try to modify columns with a complex data structure (as I mentionned, `JSONb` that's you I'm looking at 🤬).

**Comment by helsont** [comment]
Our use case: we need to create a new table in Prisma. Old users need to have this table populated by hand depending on specific conditions (like whether they are actively paying, or have more than 10 team members). New users can just get the default row. This isn't the kind of thing we'd want to write by hand in SQL. It'd be good to have this in Typescript so we can have unit tests attached to it. 

At the minimum, allow for execution of a `.ts` file in the migration folder so we can run the actual script elsewhere.

**Comment by vladinator1000** [comment]
If anyone is interested, I developed a custom seed system that lets you alter the database and fill it with data using a file-driven API. https://github.com/vladinator1000/prisma-template

**Comment by elderapo** [comment]
> If anyone is interested, I developed a custom seed system that lets you alter the database and fill it with data using a file-driven API. https://github.com/vladinator1000/prisma-template

How is this related to this issue?

**Comment by BenStirrup** [comment]
Any update on this feature, that is running migrations from JavaScript/TypeScript files ? Django's ORM has been supporting this feature for years, it is huge let down regarding data migration that Prisma cannot support.
As far as I know, TypeORM does support this feature.

**Comment by marek-hanzal** [comment]
Hello - I'm joining this with a few thougths:
- I like the OP's API proposal
- I don't like an idea Prisma Migration having an API copying schema definition - currently Prisma does great job generating SQL files, only thing it's useful could be migration script (TS/JS) itself
```
   // ***** I don't like this kind of idea as we already have Prisma Migrate SQL generator working great, no reason for doing this manually
   // and if yes, there is an ability to run raw SQL queries, so again, no need to add this kind of stuff to Prisma
    await prisma.db.addColumn(...)

    // Data migration
    await prisma.someModel.update(/*..*/)

    // Prisma migration for db cleanup
    // only once we know the data is safe in another location and defaults setup
    // can we drop columns and make columns non-nullable
    await prisma.db.alterColumn(...)
    await prisma.db.dropColumn(...)
```
- Migration script could run next to SQL, so when `prisma migration dev` is run, there could be generated SQL (if there are any database changes) AND script all the times; one not needed could be simply removed
- Default order could be SQL migration, TS/JS migration file; if there is need for another order, simply create another migration (for example do some data magick and in another migration do SQL magick)
- No "down" method; I'm not sure, how in common this "feature" is used, but mostly I see a lot of changes doing quite heavy and some times even not reversible operations, so "down" is just boilerplate written, but "never" executed nor properly tested (I suppose)
- Eventually I like @vnenkpet's idea

And now, it's time to have a look on TypeORM...

**Comment by marek-hanzal** [comment]
Ok, so I've found a kind of "solution", but I suppose a lot of you would be happy - I hope:
**The Stack**
- Prisma (yes, we all love it :) !)
- Umzug
- ts-node (and a lot of patience)
- pure TS; I've not tested pure JS as there is some pain with module loading and so
- requires `umzug`, `ts-node` and `tsconfig-paths` if you want paths mapping support in ts-node

**The Workflow**
- Prisma Migrate dev --create-only - you know it all; at this point Prisma could smell something bad around it, so it could want to reset a database; idea is to clear it manually and run Umzugs
- prepare your stuff in Prisma's native folders (/prisma/migrations/xxx/`migration.sql` + `whatever.ts`)
- when migrating, run `node --no-warnings --loader ts-node/esm -r ts-node/register src/cli/migration.ts` - whatever you put/prepare the Umzug config/entry point

**The Problem**
- Manually creating migration folders with just TS/JS code
- There is need at least an empty .sql file to prevent Prisma Migrate dev crying (only during eventual resets)
- Prisma Migrate may reset the database, so one have to manually clean up database and run Umzugs; this could be pain, but still it's some kind of solution

**The Result**
- you got mix of SQL and TS file support
- you can write native TS and execute it directly without need to transpile it with webpack and co.
- migrations are managed by Umzug, so Prisma Migrate is used just for the database diff magick
- and some pain with developing this stuff as Prisma Migrate still needs to be used and this kind of workflow could make it a bit sick

**The Umzug**
(`src/cli/migration.ts`)
```ts
import prisma from "...your prisma client...";
import {JSONStorage, Umzug} from "umzug";

export const runSql = (source: string, prisma: PrismaClient, timeout: number = 1000 * 60) => {
        // Sql.execute is just wrapper around splitting SQL file into statements and them running them through handler (in this case prisma raw stuff)
	return prisma.$transaction(async prisma => Sql.execute(source, prisma.$executeRawUnsafe), {
		timeout,
	});
};

new Umzug({
	migrations: {
		glob: process.cwd() + "/prisma/migrations/**/*.{ts,sql}",
		resolve: params => {
			const {context} = params;
			if (params.path?.endsWith(".sql")) {
				return {
					name: params.path,
					up: async () => runSql(params.path as string, context),
				};
			}
			const defaults = Umzug.defaultResolver(params);
			return {
				...defaults,
				name: params.path!!,
			};
		},
	},
	storage: new JSONStorage(),
	context: prisma,
	logger: console,
}).up()
	.then(() => process.exit(0))
	.catch(e => {
		console.error(e);
		process.exit(1);
	});
```

Maybe I forgot something, but at least it could be a starting point. I'm using Uzmug json storage to keep things simple, for example when you have already running app, you don't want Uzmug to mess everything up with already applied migrations, so you can fill the `umzug.json` file manually.

**Thing to solve** - migration name is full file path, so this should be replaced with only migration path part.

So, that's it. Happy Prisma Uzmugging!

**Comment by marek-hanzal** [comment]
Hi,

So one last thing: I made a solution compatible with Prisma Migrations:
- you can generate Prisma Migration (even with applying it)
- you can create your own TS migration (applying it by Umzug)
- migration history is saved in Prisma, so it will not break things and it's maximally compatible with Prisma, basically like it is it's feature
- there are a few limitations, but not critical; also it's not battle tested, so there could be some problems

If anybody is interested, I will create a sample repo or something like that.

See ya!

**Comment by janpio** [comment]
(That sounds like something many people in this thread would be happy about, so if you have the time, go ahead @marek-hanzal !)

**Comment by marek-hanzal** [comment]
So, this is initial snapshot:
https://github.com/marek-hanzal/prisma-migration-example

It should be a starting point, there are some references to my current project I'm using where is real-world (working) example, but I've picked up what I think is necessary.

Please give mi feedback there if you want. I hope somebody find it useful, see ya later!

**Comment by nickluger** [comment]
> (That sounds like something many people in this thread would be happy about, so if you have the time, go ahead @marek-hanzal !)

Does this mean you leave this up to (hacky) user space? Or is there already a simplified feature request without the TS DSL stuff?

For many here, it seems, it'd be sufficient to run simple `js` files instead of `sql` files, using raw SQL in migrations. Nothing fancy, like:

```js
const query = someSimpleClient.raw`
   CREATE TABLE "Foo" ( ...
`
someSimpleClient.run(query);
```

Aren't procedural migrations a necessity for any slightly more complex migration? Like e.g. splitting data from one table/column to two newly created tables/columns or just transforming data in some way etc.

> No "down" method; I'm not sure, how in common this "feature" is used, but mostly I see a lot of changes doing quite heavy and some times even not reversible operations, so "down" is just boilerplate written, but "never" executed nor properly tested (I suppose)

**=> YES 💯% true**, push another migration if you want to go "down" = revert, same as you would push another commit in your code repo.

**Comment by janpio** [comment]
> Does this mean you leave this up to (hacky) user space?

No, it means a user asked if someone was interested and I replied for all the people in this thread that there indeed are.
If we considered this done with that, we would have closed the issue.

**Comment by JulianAtPave** [comment]
Tossing in my hat to say that my org is fairly deep into Prisma in production and I'm having to hand-build a data migrations pipeline because it's missing in Prisma. I'm going to go with a similar solution to @marek-hanzal to use umzug for managing which migrations have run or not, but it seems a heck of a feature to be missing.

**Comment by justindasilva** [comment]
I was in need of this feature today and was surprised to find out that I couldn't do this in Prisma as I had been using it back in the day with Drupal (using hook updates) for the longest time.

I would love to see this officially supported by Prisma.

**Comment by leohaarmann** [comment]
### The easiest way would be to **differentiate between schema migrations and data migrations**
Some may argue that other ORMs like TypeORM do not differentiate between different types of migrations but that would always lead to type inconsistency in Prisma’s context. In addition, Redwood also took a similar approach (https://redwoodjs.com/docs/data-migrations).
When _adding a column imperative (or do anything else causing schema changes)_, the _client has to be generated to sustain typesafety_. Here is a fictional code snippet to demonstrate the problem:
```ts
prisma.user.addColumn('company');
prisma.user.create({
  data: {
    company: 'example company name', // <-- this entry does not exist yet!
    // .. other entries
  },
});
```

### The new data migration part
To create a new data migration I would suggest extending the migration command with a new **--data flag** like
`npx prisma migrate dev --data --name exampleDataMigration`

This command **creates a folder** in the default migrations folder **containing one TypeScript** file following the naming convention (unix Timestamp + migration name in snake case).

The generated TypeScript file **exports a class implementing the PrismaDataMigration** interface which provides **two functions with the prisma client as parameter**. Here is an example:
```ts
export class TimestampMigrationName implements PrismaDataMigration {

  // called when migration gets applied
  async up(prisma: PrismaClient) {
    await prisma.user.create({
      data: {
        company: 'example company name',
        // .. other entries
      }.
    });
    // more data migration logic here..
  }

  // called when migration gets reverted
  async down(prisma: PrismaClient) {
    await prisma.user.deleteMany({
      where: {
        company: 'example company name',
      },
    });
    // more revert logic here..
  }

}
```

All **data migrations should be tracked in the migration history** (_prisma_migrations table) too and executed in the default order (_oldest first, newest last_).  
Before deploying the data migration it is very important to generate the client.

**What are you thoughts on this concept?**

**Comment by elderapo** [comment]
@leohaarmann migrations should not depend on `PrismaClient` because if they do you will need to keep all the previous versions of the `PrismaClient` (with different schemas) each for every migration.

**Comment by leohaarmann** [comment]
> @leohaarmann migrations should not depend on `PrismaClient` because if they do you will need to keep all the previous versions of the `PrismaClient` (with different schemas) each for every migration.

Not necessarily. Take a look at the **long-term maintainability section from Redwood** (https://redwoodjs.com/docs/data-migrations#long-term-maintainability). 

Sure, it is a trade-off but using the PrismaClient ensures typesafety in all migrations which is more important in my opinion. Especially because one of Prisma's **biggest claims is being typesafe**.

**Comment by elderapo** [comment]
@leohaarmann What if:
- First migration creates a sample user with `username`, `password` and `email` columns.
- Second migration either deletes the `email` column or adds a new column `age`.
- First migration is no longer compilable by typescript.

It simply won't work unless you keep multiple versions of the `PrismaClient`.

**Comment by leohaarmann** [comment]
> @leohaarmann What if:
> 
> * First migration creates a sample user with `username`, `password` and `email` columns.
> * Second migration either deletes the `email` column, adds a new column `age`, or renames `username` to `name`.
> * First migration is no longer compilable by typescript.
> 
> It simply won't work unless you keep multiple versions of the `PrismaClient`.

This problem already has been discussed here and yes, **you definitely have to write some extra code** to get no TypeScript errors. That is what I meant with **trade-off**.

I think @franky47 made it quite clear here:
> @LMaxence I was more referring to "data migrations": once the schema is in a given state (changed either via the current `prisma migrate` SQL-based approach or by a code-driven approach as you propose), running code at runtime to modify records based on some logic.
> 
> This is often used in [expand-contract migrations](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern), where for example you would want to add a new column which contains data that can be computed from a mix of the existing record and external data:
> 
> 1. Add the new column, marked as optional
> 2. Run data migrations to iterate over records and fill up the new column with computed data
> 3. Mark the new column as required and remove old columns (for normalisation)
> 
> If step 2 was to use the Prisma Client to run those data migrations, then it would be hard to keep it in sync across schema versions. But because this is a TypeScript matter only, and those data migrations should only run once like their SQL counterparts, a simple `// @ts-nocheck` at the top of the data migration file would work to suppress those errors.

By the way, another trade-off would be that you can not squash data and schema migrations (https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/squashing-migrations). 

So coming to your example you would have to modify the first migration: 
As @franky47 mentioned you could simply add  `// @ts-nocheck` at the top of the data migration file.
Alternatively you could also just use type assertions like
```ts
import { Prisma } from '@prisma/client';

prisma.user.create({
  data: {
    laterDeletedColumn: 'value123',
  } as Prisma.UserCreateInput & { laterDeletedColumn: string }, // or just as any
});
```

So just to be super clear: You definitely have to **modify your TypeScript code in rather ugly way** (with type assertions) **or just ignore it** with `// @ts-nocheck` to work with the prisma client.  

But since migrations usually **only run once and in almost no one will every rewrite old (already executed) migrations**  it is a **conscious trade-off**.

I would definitely prefer having typesafety while implementing a new migration and add a few `// @ts-nocheck` to old migrations than not having this feature and write complex SQL on my own :) 


@elderapo besides this concept, how would you concept a solution for this issue?

**Comment by JulianAtPave** [comment]
This same discussion came up at my organization and we landed on a similar conclusion to @leohaarmann . You can't easily make it so that every data migration will always work forever, but who cares? 

I want the migration to work *now* because data migrations are (IMO) by definition run once. If you *need* a data migration in order for your tables to be consistent, that's not a data migration, that's a schema migration and transactional guarantees become more important than things like keeping the tables from locking or type safety.

Similarly, as @leohaarmann pointed out you actually *can* maintain those old migrations if you *really* care but just commenting them out or adding some ignores does the trick too

**Comment by elderapo** [comment]
@leohaarmann Migrations are kind of "write and forget" so I think writing/generating (`typeorm` style) raw SQL queries or using API like:
```ts
await prisma.db.alterColumn(...);
await prisma.db.dropColumn(...);
```
then making sure they work, and "forgetting about them" is the way to go.

If you want to have type-safety when writing migrations then there is a possibility of writing migrations using `PrismaClient` and after you're happy with the migration you wrote (after you made sure it's correct and it works as expected) replace calls like:
```ts
prisma.user.create({
  data: {
    username: "Tom",
    age: 123
  }
});
```

with

```ts
prisma.query(`INSERT INTO user ...`);
```

either manually or maybe using some kind of codegen.

**Comment by JulianAtPave** [comment]
>I think manually writing/generating (typeorm style) raw SQL queries or using API like:

I don't see the major benefits to this versus just doing data migrations through the existing Prisma schema migration tool? You get TS and query building but if you're already in raw query land why not just write raw queries and have them be nice and centralized in your migrations folder?

My org also ran into this *exact* problem in that nobody could nail down a sufficiently precise definition of "data migration". I interviewed a number of devs and that was the problem I kept running into that people kept placing whatever their particular immediate or non-immediate need was on top of the idea of a "data migration".

For example, does a data migration need to have:

1. Transactionality?
2. The ability to be run from any version of the schema?
3. The ability to be run from any version of the underlying app?
4. The ability to be rolled back?
5. The ability to be run more than once?
6. Any sort of built in type safety?
7. Plenty more I can't think of since I gave up on the project

People having different answers to those sorts of questions leads to wildly different opinions on what is obviously going to work or not. To me a data migration is:

1. Transactional by default, but can be overrideen
2. Is written under the assumption that there is a single target schema for the migration to run. All previous and future schemas are irrelevant to any individual data migration because:
3. data migrations are run once to fix a problem. Either they fix the problem or they don't, but if you need to run it multiple times with varying inputs - it's not a data migration.

To me, that makes using the PrismaClient in an umzug style up/down function a no-brainer. That being said, I work in an org where there is one live version of the schema at any given time. If you were designing for being able to handle many different schemas all being considered equally valid across the org, this requires additional coding to make things backwards compatible, but I don't think there's any way to get backwards compatibility for "free" other than to ignore type safety relative to the DB schema.

**Comment by elderapo** [comment]
> I don't see the major benefits to this versus just doing data migrations through the existing Prisma schema migration tool?

I don't think the existing Prisma schema migration tool works for all the migration cases that's why this issue was created. How else would you write a migration for adding a non-nullable column assuming there are already production environments in the wild? You need to:
1. Add a temporary nullable column.
2. Fix existing records so all previously inserted records (with old schema) don't have null in the newly created column anymore.
3. Alter the newly created table column to be non-nullable.

The above migration is possible to create using raw SQL queries assuming you have somewhere to pull data from (for point 2).

---

Using `PrismaClient` for migrations is a bad idea because `PrismaClient` essentially changes with every schema change (and other changes like middleware, extensions, etc.) and what that does is change the behaviors of previously written migrations which should be immutable (after it has been deployed on some important environment). For the same reason, migrations should not import "reusable" code because this "reusable" code can be changed in the future which breaks the immutability of the previous transactions.

I'd add `immutable` and `self-contained` points to your migration definition.

**Comment by franky47** [comment]
To data-migrate non-nullable new columns based on computed data (whether from the database or from outside sources), while keeping production downtime to a minimum (ideally zero), there's quite a little dance to set up.

In a previous comment, I linked to the [expand and contract pattern](https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern), which involves not only database operations, but also adapting the application code for correct reads and writes based on what phase we're at. It's not something that can be automated away with a tool.

Now, coming back to doing data migrations with Prisma Client, there are a few roadblocks in place, aside from typesafety:
1. There is no easy way to iterate on every record in a table. There are pagination utilities, but that adds a lot of boilerplate around iteration. [Streams](#5055) could help here.
2. It's likely that row migration code (computing values based on other data) will need to run in a transaction. Interactive transactions are great for this, but since they are experimental, it requires enabling them first. Not really an issue, as I found those to work well, and they'll probably stabilise soon.

**Comment by leohaarmann** [comment]
### No perfect technical solution
I think the goal of many here is to create a solution with build-in mechanisms which prevents developers from writing "bad" code. But as we discussed before - there is not one specific definition for bad code, it heavily depends on the developer you ask and the current environment (like dev team size, ...). 

@JulianAtPave made it quite clear with his statement here: 
> My org also ran into this _exact_ problem in that nobody could nail down a sufficiently precise definition of "data migration". I interviewed a number of devs and that was the problem I kept running into that people kept placing whatever their particular immediate or non-immediate need was on top of the idea of a "data migration".

_I do not think we will get all on the same page when discussing the question of good/bad code._ 
Here is a compromise which will not pursue any mechanisms concerning code quality/choices:

### Code migrations

A code migration can be created by using the following CLI command
`npx prisma migrate dev --code --name exampleCodeMigration`

It created a folder containing a TypeScript file, which implements the PrismaCodeMigration interface and follows the given naming conventions (unix Timestamp + migration name in snake case).
In contrary to my previous suggestion, the implemented interface does not pass any parameters to the up/down functions.

_It would look like this_:
```ts
export class TimestampMigrationName implements PrismaCodeMigration {

  // called when migration gets applied
  async up() {
    // do whatever you want to do here
  }

  // called when migration gets reverted
  async down() {
    // do whatever you want to do here    
  }

}
```


**No build-in mechanisms or stuff like that.**
You can implement whatever you want to - it **is up to you as developer to decide** whether it makes sense.

In order to make **data migrations** like in my previous example I would just have to **import the prisma client and run the generate command upfront** _(which is another issue mentioned here #[4703](https://github.com/prisma/prisma/issues/4703))._


By the way - right now nothing really prevents you from writing SQL code which breaks the schema and typesafety when running `npx prisma migrate dev --create-only`, so why shouldn't it be possible to have the same behavior in TypeScript?

### co-existence
At then end both ways (_new code migration and raw sql only_) would just co-exist. The raw SQL approach already exists with `npx prisma migrate dev --create-only`. Devs who prefer @elderapo's approach could still go with raw SQL. Devs like me would go with the new code migrations.

### Easy maintenance and competitors
In addition, this feature would be super easy to maintain since it does not implement any constrains. 
Other ORMs like TypeORM, Sequelize, Django's ORM and ActiveRecord offer running code in migrations. As mentioned in a few comments before like @BenStirrup's comment, it is crucial for many dev teams especially in production environments. 
> Any update on this feature, that is running migrations from JavaScript/TypeScript files ? Django's ORM has been supporting this feature for years, it is huge let down regarding data migration that Prisma cannot support. As far as I know, TypeORM does support this feature.
> 
> **Edit** : we eventually decided to use TypeORM instead of Prisma, and the lack of this feature was a huge point in deciding to do so...


**What are your thoughts on the code migration approach?**

**Comment by janpio** [comment]
Quick note from the Prisma side (who is obviously watching this conversation): The most important part of the issue here is you all leaving a precise description of your use case. As you also highlighted "data migrations" means all kinds of things to different people, so having an as extensive as possible list of these is good start for us. The design suggestions are of course also welcome, but the use case description are _essential_ to use to be able to figure this out later. Thanks.

**Comment by kokokenada** [comment]
My use case for this is that I use Postgres JSONB field types to sweep up complex data structures. There's a typescript interfaces that describes the format (and often helper/utility functions).  From time to time I need to change the format of the JSON and want to run a migration.  This is possible  at the database level with postgres JSON support but much more painful compared to writing something in typescript where I can leverage the types and other code that I have.   So as part of running the migration, I want to be able to import code from my app.

**Comment by villesau** [comment]
I have a _very_ simple use case: I need to be able to generate CUIDs while splitting a table and populating data from old table to the new one. I don't want to copy the id from table to another.

I think the default and the first class method could still very well be an sql file, but a possibility to run raw sql mixed with JS would be helpful in many more complex cases.

**Comment by antmarot** [comment]
My use case is a column containing some kind of instructions (in a given DSL) as strings.
We want to drop the support of that DSL in favor of JSON blobs, which will make it easier to integrate with for consumers.
I would need the ability to run code in order to transform the former format into the new one in a new column.

**Comment by cyrilchapon** [comment]
I basically don't understand any argument here, nor am able to find a legit one against improving migrations with a DSL language for this.

Prisma is a fantastic piece of library, such a wonderful ORM. But everytime I ask myself "which ORM should I choose this time ?" (On every project I start), **every. single. time** the migration topic is in the "cons" column for Prisma.

Almost 400 issues with a title containing the word "migration" should be a strong hint that it's time to face the reality : it was a nice try, but the migration system is bad. As simple as that.

The above comment states that "what needs the Prisma teams are some piece of use-cases"; well here it is, as a user-story :

> As a developer, I want to generate migrations for my database, with an understandable CLI, then tweak the migrations to fit my needs, in a language which is not from the 90s

1. That distinction between "structure" and "data" migration...

...is a total nonsense. In the actual world, when you migrate the structure, 50% of the time you'll migrate the data.

The only correct way to do this, is to do this in a transaction. Oh wait... not every DB systems can use migration for data structure ? Guess what : you just have to tell them to switch to a non-crappy DB engine !

Examples :
- split a table in 3 (create / select / insert / drop)
- change the id column type (Alter / update / alter)

2. PrismaClient cannot be used here because it changes when generated

Well refactor it. Underlying low level proxy methods should just not be part of it; and should not change when generated.

3. We don't want to encourage people to write bad code

Yeah for sure. Let's rather encourage them writing SQL by hand, and bypassing the migration flow when migrating data outside of migration transactions.

Total nonsense.

---------

This might sound spicy, and I'm sorry about that.  But please take note that I LOVE that ORM; deeply respect the hard work done on it. And all of this are just IMHO legit arguments from someone (just like most backend developers) who has written hundreds of migrations over the years, have worked with tens of ORMs on tens of environments, and is dying to finally be able to use this fantastically promising ORM without thinking "ha, I had forgotten that strange migration system"...

**Comment by nickluger** [comment]
> Yeah for sure. Let's rather encourage them writing SQL by hand, and bypassing the migration flow when migrating data outside of migration transactions.

With Prisma I started living the thug life and runnin ma `run-manual-migration.ts` script whenever I want 😎. CI can't stop me!

**Comment by emmanuelbuah** [comment]
I have kept a close watch on this topic and aware the prisma community cares a lot about it. Db migration is a difficult problem to solve but the cleanest I have seen is with edgedb - https://www.edgedb.com/docs/intro/migrations. It not perfect but it's the best I've see so far.

I use prisma and hope it gets inspired by edgedb in improving its migration strategy.

**Comment by cyrilchapon** [comment]
@emmanuelbuah yes, and the same goes with [umzug](https://github.com/sequelize/umzug); when used programmatically which can become a solid brick for simple and powerful migration systems too.

Here is an example from a personal repository :

#### README.md

(notice how things are simple-and-stupid, without any magic in it)

```markdown
yarn mig --help

# Execute all pending migrations
yarn mig up

# Execute next migration
yarn mig up --one

# Revert all migrations
yarn mig down

# Revert last migration
yarn mig down --one

# Get migrations status
# Success if up-to-date
# Errors if pending found
yarn mig status

```

#### migrator.ts
```typescript
import { Command } from '@commander-js/extra-typings'
import { appEnv } from '../lib/env/app-env.js'
import { createWhateverDbConnector } from '../lib/whatever-db-connector.js'
import { createUmzug } from './umzug.js'

const whateverDbConnector = createWhateverDbConnector(appEnv.DATABASE_URL)
const umzug = createUmzug(whateverDbConnector, 'migration')

const program = new Command()
program.name('mig').description('Perform database migrations powered by Umzug')

program
  .command('create')
  .description('Create a new migration')
  .argument('<name>', 'name of migration to create')
  .action(async (name) => {
    await umzug.create({
      name: name,
      prefix: 'NONE',
      allowExtension: '.ts',
      allowConfusingOrdering: false,
      skipVerify: false,
    })
  })

program
  .command('up')
  .description('Execute pending migrations')
  .option('--one', 'just execute next migration')
  .action(async (options) => {
    await umzug.up(!!options.one ? { step: 1 } : undefined)
  })

program
  .command('down')
  .description('Revert executed migrations')
  .option('--one', 'just revert previous migration')
  .action(async (options) => {
    await umzug.down(!!options.one ? undefined : { to: 0 })
  })

program
  .command('status')
  .description('Get migrations status')
  .action(async () => {
    const pendingMigrations = await umzug.pending()
    if (pendingMigrations.length > 0) {
      program.error(
        `Found pending migrations, execute \`mig up\` to run them.\n${JSON.stringify(
          pendingMigrations.map((mig) => mig.name),
        )}`,
        {
          exitCode: 1,
          code: 'pending.migrations',
        },
      )
    }
  })

const go = async () => {
  try {
    await program.parseAsync()
  } finally {
    await mongo.close()
  }
}

void go()
```

migration-xxx.ts
```typescript
import type { Migration } from '../umzug.js'

export const up: Migration = async ({ context: { whateverDbConnector } }) => {
  await whateverDbConnector.transaction(async (tx) => {
    await tx.createTable('stuff')
    // await some data migration
    await tx.dropTable('other-stuff')
  })
}

export const down: Migration = async ({ context: { whateverDbConnector } }) => {
  await whateverDbConnector.transaction(async (tx) => {
    await tx.createTable('other-stuff')
    // await revert data migration
    await tx.dropTable('stuff')
  })
}
```

Things are so simple here, understandable, flexible. What is the point of missing that ?

**Comment by Daidalos117** [comment]
I was trying one of the ways with umzug. After using it, the Prisma migration tool (when I want to create new SQL migration) detects drift. So basically it's not possible to use umzug and prisma migration together right?
Anyway, it's really really bad that Prisma itself doesn't solve this by itself. I honestly cannot comprehend how any bigger application can work without script migrations.

**Comment by cyrilchapon** [comment]
@janpio , that's so great to hear ❤️

If I could add just one thing here, is I also tried myself to just-don't-use-prisma-migration-tool and go with another one like Umzug; and faced the same issue.

It would be a great addition, maybe, to allow Prisma just to report the DB "state" versus Schema. Just that; and print this to the console; maybe with an option to do this in CIs.

**Comment by kodmanyagha** [comment]
I'm using prisma for db operations, typeorm for migrations. Prisma's migration feature is a HUGE SHIT.

**Comment by juliavdkris** [comment]
@kodmanyagha hey I'd also like to see prisma's migrations improve, but let's stay respectful to open-source developers who work on these projects in their free time.

**Comment by kodmanyagha** [comment]
Yes, you're right. I'm respecting to all opensource or closedsource developers, I didn't say anything to them. I'm talking about only prisma library itself.

**Comment by juliavdkris** [comment]
No, you're talking shit about something that people pour a lot of time and energy in for free, in a way that is not at all constructive. You're adding nothing to the conversation other than unnecessary negativity.

If you wanna improve prisma's migrations, feel free to make a PR lmao

**Comment by cyrilchapon** [comment]
@juliavdkris are we gonna argue around an angry word on a 4-years-old non moving and critical issue ? Let's go then, how constructive. So many cognitive bias on so few lines is a great performance, congrats.

Firstly, the "disrespect to open source developers" working on this project does not come from an angry issue comment, but rather from the nonsense opiniated governance forcing them to maintain a 3k-issues-features that nobody in the community likes. "Open source" has spoken, by raising tons of issues on those migrations, and we're not talking about their IMPLEMENTATION (developer job) but rather about their DESIGN (not developer job).

Secondly, the fact Prisma is now a fundraised organization with a governance, commenting one time a year on a 4-years-old, well documented, well starred, critical issue, is precisely the definition of "disrespect", not only to open source developers pouring hard work using that library, but to anyone having a single interest in this. Including internal developers themselves.

Lastly, avoiding true words and refusing constructive criticism just like you just did is actually not gonna engage people to "make a PR lmao"; but rather find an alternative library and give up on this nonsense governed one, including talented developers who would have been able to "make some PRs lmao" elsewhere in the stack.

So yeah, @kodmanyagha, don't "make a PR lmao", but rather go with Drizzle lmao. That will be a much stronger signal for this governance to tackle this, actually truely bad, migration system.

**Comment by erikc96** [comment]
It looks like Prisma supports this feature now anyway?

https://www.prisma.io/docs/orm/prisma-migrate/workflows/data-migration

```
Create a data migration file

Inside the generated migration folder from the previous step, create a file called data-migration.ts file. This file will contain a data migration which will be implemented using Prisma Client.

Add the following code to migrate the data from the published field to the status field in the file you just created:

Update your package.json file to include the data migration file:
prisma/migrations/20230417131956_add-status-column/data-migration.ts

import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

async function main() {
  await prisma.$transaction(async (tx) => {
    const posts = await tx.post.findMany()
    for (const post of posts) {
      await tx.post.update({
        where: { id: post.id },
        data: {
          status: post.published ? 'Published' : 'Unknown',
        },
      })
    }
  })
}

main()
  .catch(async (e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => await prisma.$disconnect())

The data migration is wrapped in a transaction to ensure that the query is rolled back, allowing you to iterate on your data migration file
```

**Comment by tmax22** [comment]
@erikc96 no Prisma does not support code migrations yet. they just added a small tutorial that demonstrate how to execute data migration **manually** which is super fragile in my opinion because this means that the developer must remember to apply the migrations in the right order. I really don't get it, Prisma is a fantastic ORM but its migration system are just very limited, and 4 years have passed with little to no change on this feature. **Prisma team, please support code migrations already**.

**Comment by sebasgarcep** [comment]
After waiting long enough for a fix I decided to hack a migration system that could handle both SQL and TS files. Here's the gist for anyone interested: https://gist.github.com/sebasgarcep/6af4e6b63dc1543bbcafa2f2676f3312. This is how it works:

```sh
# Automatically generate SQL using prisma migrate dev under the hood
pnpm run migrate create <migration name>

# Create an empty TS migration
ppm run migrate create <migration name> --type=ts

# Apply migrations
pnpm run migrate up
pnpm run migrate up --to=<timestamp_migration name>

# Roll back migrations
pnpm run migrate down
pnpm run migrate down --to=<timestamp_migration name>

# Check if there are unapplied migrations
pnpm run migrate status
```

It only works with Postgres and the TS migrations get a `pg.Client` instead of a `PrismaClient` because it would've been more work to support transactional DDL otherwise. You also need to set up a shadow database url, otherwise it won't work properly.

Thanks to @cyrilchapon for his Umzug sample. It was a fantastic starting point! ❤️

**Comment by ekr3peeK** [comment]
For my use-case, even a simple hook script, that can be ran after a specific migration would do

**Comment by Softjey** [comment]
> For my use-case, even a simple hook script, that can be ran after a specific migration would do

Hi everyone! 👋

I’ve developed a small library to help address this problem: [Prisma Data Migrations](https://www.npmjs.com/package/prisma-data-migrations?activeTab=readme). This library enhances Prisma’s migration system by enabling the use of post-migration scripts for data migrations alongside schema changes.

You can find the repository here: [GitHub Repository](https://github.com/Softjey/prisma-data-migrations).

Feel free to open issues, report bugs, or suggest improvements. I’m also open to pull requests if you’d like to contribute to the project!

Looking forward to your feedback! 🚀

**Comment by karlhorky** [comment]
I can also contribute my simplistic fixtures seeding script (using Postgres.js, but same principles would apply to creating such a script with Prisma), in case this can help anyone:

- https://github.com/karlhorky/postgresql-tricks#seed-test-fixture-data-with-explicit-ids-to-generated-identity-fields

The order of the insertions is handled by numeric prefixes in filenames, similar to the common practice in migration files.

It also allows for inserting explicit IDs in identity columns (by temporarily disabling the identity column and then adding back), in case your test data can benefit from having explicit IDs.

**Comment by tsito-irok** [comment]
Almost 5 years later, nothing yet ?

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319
